tools/cache_warm.cc

```diff
@@ -0,0 +1,280 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+// -------------------------------------------------------------------
+
+#include <errno.h>
+#include <libgen.h>
+#include <limits.h>
+#include <memory>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "include/leveldb/env.h"
+#include "db/dbformat.h"
+#include "db/filename.h"
+#include "db/log_writer.h"
+#include "db/version_edit.h"
+#include "util/coding.h"
+
+void command_help();
+
+// wrapper class for opening and writing a new Cache Object Warming file
+class CowFile
+{
+public:
+    CowFile(std::string & DatabasePath, std::string & Record);
+    virtual ~CowFile() {};
+
+    const leveldb::Status & GetStatus() const {return(m_LastStatus);};
+    const char * GetFileName() const {return(m_FileName.c_str());};
+
+protected:
+    std::string m_FileName;
+    leveldb::Status m_LastStatus;
+
+private:
+    // disable these
+    CowFile();
+    CowFile(const CowFile &);
+    const CowFile operator=(const CowFile&);
+};  // class CowFile
+
+
+CowFile::CowFile(
+    std::string & DatabasePath,
+    std::string & Record)
+{
+    leveldb::WritableFile * cow_file;
+    leveldb::log::Writer * cow_log;
+
+    m_FileName=leveldb::CowFileName(DatabasePath);
+    m_LastStatus = leveldb::Env::Default()->NewWritableFile(m_FileName, &cow_file, 4*1024L);
+
+    if (m_LastStatus.ok())
+    {
+        cow_log=new leveldb::log::Writer(cow_file);
+
+        m_LastStatus = cow_log->AddRecord(Record);
+        delete cow_log;
+        delete cow_file;
+    }   // if
+
+    return;
+
+}   // CowFile::CowFile
+
+
+int
+main(
+    int argc,
+    char ** argv)
+{
+    bool error_seen, running;
+    char ** cursor;
+
+    error_seen=false;
+    running=true;
+
+
+    for (cursor=argv+1;
+         NULL!=*cursor && running && !error_seen;
+         ++cursor)
+    {
+        // option flag?
+        if ('-'==**cursor)
+        {
+            char flag;
+
+            flag=*((*cursor)+1);
+            switch(flag)
+            {
+                default:
+                    fprintf(stderr, " option \'%c\' is not valid\n", flag);
+                    command_help();
+                    running=false;
+                    error_seen=true;
+                    break;
+            }   // switch
+        }   // if
+
+        // file of .sst pathnames
+        else
+        {
+            FILE * list;
+            bool ignore_file;
+            std::string cur_path, record;
+            size_t line_count;
+            char buffer[PATH_MAX], vnode[PATH_MAX], file[NAME_MAX], * chr_ptr;
```

Comment:
Paul Place writes: NAME_MAX does not include the terminating null character, so should make it file[NAME_MAX+1]